cdap/http.go

```diff
@@ -37,13 +38,26 @@ func urlJoin(base string, paths ...string) string {
 	return fmt.Sprintf("%s/%s", strings.TrimRight(base, "/"), strings.TrimLeft(p, "/"))
 }
 
-func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
-	log.Printf("%+v", req)
-
+func do(client *http.Client, req *http.Request) (*http.Response, error) {
 	resp, err := client.Do(req)
 	if err != nil {
 		return nil, err
 	}
+	return resp, nil
+}
+
+func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
+	log.Printf("%+v", req)
+
+	resp, err := do(client, req)
+
+	// CDAP REST intermittently returns 500 internal errors, we will retry on 500s once.
+	if resp.StatusCode == 500 {
```

Comment:
https://golang.org/pkg/net/http/#Client.Do

> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

So I think you want to actually move the entire logic that reads the status code into the do func

Reply:
Not sure I follow why this belongs in the (newly renamed) `doHttpCall` func.
I thought resp code should be read here because this is how I retry only once (by calling `doHttpCall` again rather than doing some sort of recursion).

Reply:
Sorry, I wasn't clear. 

The code below this reads the response body and then checks for an invalid error and returns the httpError. You want that entire path to be called on every client.Do call so that the response body is fully depleted and closed.

Right now you are doing

resp, err := do()
// check 500
resp, err = do()

This isn't reading and closing the first call's response body as is suggested in the func doc.

You instead wanna do:

b, err := do()
var e *httpError
if errors.As(err, &e) {
   time.Sleep()
  return do()
} else {
  return b, err
}

where do is doing the call and reading the body and all.

Reply:
Thanks for clarification. 
this requires changing the signature of doHttpCall but makes sense, good catch.